lib/puppet/util/windows/monkey_patches/dir.rb

```diff
@@ -0,0 +1,409 @@
+    "FAVORITES",                0x0006,
+    "STARTUP",                  0x0007,
+    "RECENT",                   0x0008,
+    "SENDTO",                   0x0009,
+    "BITBUCKET",                0x000a,
+    "STARTMENU",                0x000b,
+    "MYDOCUMENTS",              0x000c,
+    "MYMUSIC",                  0x000d,
+    "MYVIDEO",                  0x000e,
+    "DESKTOPDIRECTORY",         0x0010,
+    "DRIVES",                   0x0011,
+    "NETWORK",                  0x0012,
+    "NETHOOD",                  0x0013,
+    "FONTS",                    0x0014,
+    "TEMPLATES",                0x0015,
+    "COMMON_STARTMENU",         0x0016,
+    "COMMON_PROGRAMS",          0X0017,
+    "COMMON_STARTUP",           0x0018,
+    "COMMON_FAVORITES",         0x001f,
+    "COMMON_DESKTOPDIRECTORY",  0x0019,
+    "APPDATA",                  0x001a,
+    "PRINTHOOD",                0x001b,
+    "LOCAL_APPDATA",            0x001c,
+    "ALTSTARTUP",               0x001d,
+    "COMMON_ALTSTARTUP",        0x001e,
+    "INTERNET_CACHE",           0x0020,
+    "COOKIES",                  0x0021,
+    "HISTORY",                  0x0022,
+    "COMMON_APPDATA",           0x0023,
+    "WINDOWS",                  0x0024,
+    "SYSTEM",                   0x0025,
+    "PROGRAM_FILES",            0x0026,
+    "MYPICTURES",               0x0027,
+    "PROFILE",                  0x0028,
+    "SYSTEMX86",                0x0029,
+    "PROGRAM_FILESX86",         0x002a,
+    "PROGRAM_FILES_COMMON",     0x002b,
+    "PROGRAM_FILES_COMMONX86",  0x002c,
+    "COMMON_TEMPLATES",         0x002d,
+    "COMMON_DOCUMENTS",         0x002e,
+    "CONNECTIONS",              0x0031,
+    "COMMON_MUSIC",             0x0035,
+    "COMMON_PICTURES",          0x0036,
+    "COMMON_VIDEO",             0x0037,
+    "RESOURCES",                0x0038,
+    "RESOURCES_LOCALIZED",      0x0039,
+    "COMMON_OEM_LINKS",         0x003a,
+    "CDBURN_AREA",              0x003b,
+    "COMMON_ADMINTOOLS",        0x002f,
+    "ADMINTOOLS",               0x0030
+  ]
+
+  # Dynamically set each of the CSIDL constants
+  csidl.each do |key, value|
+    buf  = 0.chr * 1024
+    path = nil
+    buf.encode!("UTF-16LE")
+
+    if SHGetFolderPathW(0, value, 0, 0, buf) == 0 # Current path
+      path = buf.strip
+    elsif SHGetFolderPathW(0, value, 0, 1, buf) == 0 # Default path
+      path = buf.strip
+    else
+      ptr   = FFI::MemoryPointer.new(:long)
+      info  = SHFILEINFO.new
+      flags = SHGFI_DISPLAYNAME | SHGFI_PIDL
+
+      if SHGetFolderLocation(0, value, 0, 0, ptr) == 0
+        if SHGetFileInfo(ptr.read_long, 0, info, info.size, flags) != 0
+          path = info[:szDisplayName].to_s
+          path.force_encoding(Encoding.default_external)
+        end
+      end
+    end
+
+    begin
+      Dir.const_set(key, path.encode(Encoding.default_external)) if path
+    rescue Encoding::UndefinedConversionError
+      Dir.const_set(key, path.encode("UTF-8")) if path
```

Comment:
Would be better to use `Encoding::UTF_8` constants. Same for other places where this is used.